fix(49537): remove position relative from echarts tooltip container

npretto opened the pull request:
Fixes #49537

See linked issue and https://metaboat.slack.com/archives/C063Q3F1HPF/p1731492251517089?thread_ts=1731311131.911449&cid=C063Q3F1HPF for the investigation done

I've commented inline.

# Demo


https://github.com/user-attachments/assets/2d803295-178f-40b5-8ca0-26b5405e9b62



# TODO
- [ ] ~tests: I'll add them if this solution is approved~ I've added tests, but `should("be.visible")` doesn't work, so I have no way to test this.

npretto committed:
attempt to fix 49537 by removing position relative from the tooltip
container

WiNloSt committed:
Merge branch 'master' into npretto-fix-49537-attempt

WiNloSt committed:
Make echart tooltip container cover all iframe area

WiNloSt committed:
Update the E2E test

WiNloSt commented on e2e/test/scenarios/embedding/embedding-dashboard.cy.spec.js:
```diff
@@ -938,10 +938,20 @@ describeEE("scenarios > embedding > dashboard appearance", () => {
+      name: "Line chart",
       query: {
         "source-table": ORDERS_ID,
+        aggregation: [["count"]],
```
I updated an existing tests, so we don't need to have a new setup which takes more time to run the test.

WiNloSt commented on e2e/test/scenarios/embedding/embedding-dashboard.cy.spec.js:
```diff
@@ -977,7 +987,25 @@ describeEE("scenarios > embedding > dashboard appearance", () => {
+      // (metabase#49537)
+      // TODO: Enable these assertions once `echartsTooltip().should("be.visible")` fails when the tooltip is not visible
+      // chartPathWithFillColor("#509EE3").last().realHover();
+      // echartsTooltip().should("be.visible");
```
This might be one of the rare cases we can't use `be.visible` since Cypress passes this assertion when the tooltip isn't visible.
![Screenshot 2024-11-14 at 5 19 26 PM](https://github.com/user-attachments/assets/7e75967a-75e6-443e-ab6d-78132ec4ad50)

I also tried [`Cypress.dom.isVisible()`](https://docs.cypress.io/api/cypress-api/dom#Is-visible) but it also doesn't fail this assertion. I suspect `should("be.visible")` uses `Cypress.dom.isVisible()` under the hood

I commented it out because this assertion still pass even when I revert the fix.

WiNloSt commented on frontend/src/metabase/visualizations/echarts/tooltip/index.tsx:
```diff
@@ -72,8 +72,8 @@ export const getTooltipBaseOption = (
         container.style.setProperty("overflow", "hidden");
-        container.style.setProperty("height", "100%");
-        container.style.setProperty("position", "relative");
+        container.style.setProperty("position", "absolute");
```
The problem is that when we use iframe resizer we set `EmbedFrame` to `position: absolute; inset: 0;`

https://github.com/metabase/metabase/blob/27a91e64fcbf4e6e43f23c18b3d207bf258b1334/frontend/src/metabase/public/components/EmbedFrame/EmbedFrame.styled.tsx#L23-L31

This makes the `#root` div have a height greater than 0. When we don't use the iframe resizer this `#root` div has no height, so this echart tooltip container sit at the top of the iframe.

See this [message on Slack](https://metaboat.slack.com/archives/C063Q3F1HPF/p1731492251517089?thread_ts=1731311131.911449&cid=C063Q3F1HPF) for more details.

I fix the issue by ensuring the echart tooltip covers the entire iframe. 

![image](https://github.com/user-attachments/assets/ce8d0b1a-38d3-4d0c-8fe9-48d4dd80bcbc)

Compare to the version before this fix
![image](https://github.com/user-attachments/assets/795cc95a-705e-4101-8edc-a8a90f8b9744)

WiNloSt committed:
Update loki snapshots from CI

WiNloSt commented on .loki/reference/chrome_laptop_embed_PublicOrEmbeddedDashboardView_Dark_Theme_Default.png:
All the snapshots have been taken [from CI](https://github.com/metabase/metabase/actions/runs/11835265967). [Download it here](https://github.com/metabase/metabase/actions/runs/11835265967/artifacts/2186928760).

If you go to this commit https://github.com/metabase/metabase/commit/c6141aa7472003576a0e91bf936318971be965b8, you'd see that all the snapshot width has shrunk from 1366px to 1351px because there was an extra scrollbar when it's not supposed to be.

![image](https://github.com/user-attachments/assets/7b30d98e-5907-4e23-8a34-d6868f1c1ab0)

The CSS in this PR also fixes that issue. This would likely to only caused problem on embedding in iframes.

deniskaber reviewed (approved):
Nice one! Really good that you found a solution for this issue

WiNloSt committed:
Merge branch 'master' into npretto-fix-49537-attempt

WiNloSt committed:
Re-enable echarts tooltip assertion.

I couldn't find a better way to fail these assertions on the previous
version of the code, but these assertions are better than nothing.

WiNloSt commented on e2e/test/scenarios/embedding/embedding-dashboard.cy.spec.js:
```diff
@@ -977,7 +990,24 @@ describeEE("scenarios > embedding > dashboard appearance", () => {
+      cy.findByText("April 2022").should("exist");
+
+      // (metabase#49537)
+      chartPathWithFillColor("#509EE3").last().realHover();
```
I re-enable these assertions following [Sasha's suggestion](https://metaboat.slack.com/archives/C064QMXEV9N/p1731615004983839?thread_ts=1731580671.819259&cid=C064QMXEV9N). I've tried a couple of other variations to fail the previous version of the code, but it still couldn't.

One thing I found on master that is different from the actual embed is that the tooltip will be located at the actual location on master, but not on Cypress. Because of this `echartsTooltip().should("be.visible")` is true which is correct.

#### Cypress
![Screenshot 2024-11-15 at 2 02 54 PM](https://github.com/user-attachments/assets/6804df2d-1bbc-47d0-81f7-1da0159a5776)


#### Actual embed
![Screenshot 2024-11-15 at 2 04 23 PM](https://github.com/user-attachments/assets/cc9cdcb9-d9e4-4d69-bc28-2ec18715df97)
